File: app/src/main/java/ifteam/affogatoman/fontgen/MainActivity.java
package ifteam.affogatoman.fontgen;

import android.app.Activity;
import android.app.AlertDialog.Builder;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Bitmap.CompressFormat;
import android.graphics.Bitmap.Config;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import ar.com.daidalos.afiledialog.FileChooserDialog;
import ar.com.daidalos.afiledialog.FileChooserDialog.OnFileSelectedListener;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;

public class MainActivity extends Activity {
    public static final String TAG = "AFFO";
    Bitmap base = Bitmap.createBitmap(256, 256, Config.ARGB_8888);
    Canvas canvas;
    Button find;
    int h;
    MyHandler handler;
    Button make;
    Paint p;
    EditText path;
    ProgressBar pb;
    ProgressDialog pd;
    int presize;
    int prex;
    int prey;
    ImageView screen;
    Button setting;
    int size;
    SeekBar sizes;
    Typeface tf;
    int w;
    SeekBar xs;
    SeekBar ys;

    /* renamed from: ifteam.affogatoman.fontgen.MainActivity$100000002 */
    static class AnonymousClass100000002 implements OnSeekBarChangeListener {
        private final MainActivity this$0;
        private final ImageView val$screen;

        AnonymousClass100000002(MainActivity mainActivity, ImageView imageView) {
            AnonymousClass100000002 anonymousClass100000002 = this;
            this.this$0 = mainActivity;
            this.val$screen = imageView;
        }

        static MainActivity access$0(AnonymousClass100000002 anonymousClass100000002) {
            return anonymousClass100000002.this$0;
        }

        public void onStartTrackingTouch(SeekBar seekBar) {
        }

        public void onProgressChanged(SeekBar seekBar, int i, boolean z) {
            this.val$screen.setImageBitmap(this.this$0.getSample(this.this$0.prex, this.this$0.prey + 13, this.this$0.presize));
            switch (seekBar.getId()) {
                case R.id.sizes:
                    this$0.presize = i;
                    return;
                case R.id.xs:
                    this$0.prex = i - 16;
                    return;
                case R.id.ys:
                    this$0.prey = i - 16;
                    return;
                default:
                    return;
            }
        }

        public void onStopTrackingTouch(SeekBar seekBar) {
            this.val$screen.setImageBitmap(this.this$0.getSample(this.this$0.prex, this.this$0.prey + 13, this.this$0.presize));
        }
    }

    /* renamed from: ifteam.affogatoman.fontgen.MainActivity$100000003 */
    static class AnonymousClass100000003 implements OnClickListener {
        private final MainActivity this$0;
        private final ImageView val$screen;

        AnonymousClass100000003(MainActivity mainActivity, ImageView imageView) {
            AnonymousClass100000003 anonymousClass100000003 = this;
            this.this$0 = mainActivity;
            this.val$screen = imageView;
        }

        static MainActivity access$0(AnonymousClass100000003 anonymousClass100000003) {
            return anonymousClass100000003.this$0;
        }

        public void onClick(View view) {
            this.val$screen.setImageBitmap(this.this$0.getSample(this.this$0.prex, this.this$0.prey + 13, this.this$0.presize));
            MainActivity mainActivity;
            switch (view.getId()) {
                case R.id.sizem:
                    if (this$0.sizes.getProgress() != 0) {
                        mainActivity = this$0;
                        mainActivity.presize--;
                        this$0.sizes.incrementProgressBy(-1);
                    }
                    return;
                case R.id.sizep:
                    if (this$0.sizes.getProgress() != 16) {
                        mainActivity = this$0;
                        mainActivity.presize++;
                        this$0.sizes.incrementProgressBy(1);
                    }
                    return;
                case R.id.xm:
                    if (this$0.xs.getProgress() != 0) {
                        mainActivity = this$0;
                        mainActivity.prex--;
                        this$0.xs.incrementProgressBy(-1);
                    }
                    return;
                case R.id.xp:
                    if (this$0.xs.getProgress() != 32) {
                        mainActivity = this$0;
                        mainActivity.prex++;
                        this$0.xs.incrementProgressBy(1);
                    }
                    return;
                case R.id.ym:
                    if (this$0.ys.getProgress() != 0) {
                        mainActivity = this$0;
                        mainActivity.prey--;
                        this$0.ys.incrementProgressBy(-1);
                    }
                    return;
                case R.id.yp:
                    if (this$0.ys.getProgress() != 32) {
                        mainActivity = this$0;
                        mainActivity.prey++;
                        this$0.ys.incrementProgressBy(1);
                    }
                    return;
                default:
            }
        }
    }

    /* renamed from: ifteam.affogatoman.fontgen.MainActivity$100000005 */
    static class AnonymousClass100000005 implements Runnable {
        long current = System.currentTimeMillis();
        int glyph = 172;
        private final MainActivity this$0;
        private final int val$size;

        AnonymousClass100000005(MainActivity mainActivity, int i) {
            AnonymousClass100000005 anonymousClass100000005 = this;
            this.this$0 = mainActivity;
            this.val$size = i;
        }

        static MainActivity access$0(AnonymousClass100000005 anonymousClass100000005) {
            return anonymousClass100000005.this$0;
        }

        public void run() {
            StringBuffer stringBuffer;
            StringBuffer stringBuffer2;
            try {
                this.this$0.base = Bitmap.createBitmap(256, 256, Config.ARGB_8888);
                MainActivity mainActivity = this$0;
                mainActivity.canvas = new Canvas(this$0.base);
                if (this$0.tf != null) {
                    Typeface typeface = this$0.p.setTypeface(this$0.tf);
                }
                this$0.p.setColor(-1);
                this$0.p.setTextSize((float) val$size);
                this$0.p.setAntiAlias(true);
                int i = 44032;
                while (i <= 55203) {
                    File file;
                    StringBuffer stringBuffer3;
                    StringBuffer stringBuffer4;
                    File file2;
                    File file3;
                    Thread.sleep((long) 1);
                    Canvas canvas3 = this$0.canvas;
                    stringBuffer = new StringBuffer();
                    canvas3.drawText(stringBuffer.append((char) i).append("").toString(), (float) this$0.w, (float) this$0.h, this$0.p);
                    MainActivity mainActivity2 = this$0;
                    mainActivity2.w += 16;
                    MyHandler myHandler = this$0.handler;
                    MyHandler myHandler2 = this$0.handler;
                    Integer num = i;
                    boolean sendMessage = myHandler.sendMessage(myHandler2.obtainMessage(100, num));
                    if (i == 44033) {
                        stringBuffer2 = new StringBuffer();
                        stringBuffer3 = new StringBuffer();
                        stringBuffer4 = new StringBuffer();
                        file = new File(stringBuffer2.append(stringBuffer3.append(stringBuffer4.append(Environment.getExternalStorageDirectory()).append("/아포카토맨/FontGen/").toString()).append(current).toString()).append("/.nomedia").toString());
                        file3 = file;
                        file3.getParentFile().mkdirs();
                        file3.createNewFile();
                    }
                    if ((i - 44032) % 16 == 15 && i != 44032) {
                        this$0.w = 0;
                        mainActivity2 = this$0;
                        mainActivity2.h += 16;
                    }
                    if (this$0.h > 254 || i == 55203) {
                        stringBuffer2 = new StringBuffer();
                        stringBuffer3 = new StringBuffer();
                        stringBuffer4 = new StringBuffer();
                        StringBuilder stringBuffer5 = new StringBuilder();
                        StringBuilder stringBuffer6 = new StringBuilder();
                        file = new File(stringBuffer2.append(stringBuffer3.append(stringBuffer4.append(stringBuffer5.append(stringBuffer6.append(Environment.getExternalStorageDirectory()).append("/아포카토맨/FontGen/").toString()).append(current).toString()).append("/glyph_").toString()).append(Integer.toHexString(glyph).toUpperCase()).toString()).append(".png").toString());
                        file3 = file;
                        file3.getParentFile().mkdirs();
                        OutputStream fileOutputStream = new FileOutputStream(file3);
                        OutputStream bufferedOutputStream = new BufferedOutputStream(fileOutputStream);
                        this$0.base.compress(CompressFormat.PNG, 100, bufferedOutputStream);
                        bufferedOutputStream.close();
                        fileOutputStream.close();
                        if (i != 55203) {
                            this$0.base.eraseColor(0);
                        }
                        mainActivity = this$0;
                        mainActivity.h = mainActivity.prey + 13;
                        glyph++;
                    }
                    i++;
                }
            } catch (Exception e) {
                stringBuffer = new StringBuffer();
                int i2 = Log.i(MainActivity.TAG, stringBuffer.append(e).append("").toString());
            }
        }
    }

    class MyHandler extends Handler {
        @Override
        public void handleMessage(Message message) {
            if (message.what == 100) {
                screen.setImageBitmap(base);
                ProgressDialog progressDialog = pd;
                StringBuilder stringBuffer = new StringBuilder();
                progressDialog.setMessage(stringBuffer.append("현재 글자 : ").append((char) ((Integer) message.obj).intValue()).toString());
                pb.incrementProgressBy(1);
                pd.incrementProgressBy(1);
                if ((Integer) message.obj == 55203) {
                    make.setClickable(true);
                    setting.setClickable(true);
                    pb.setProgress(0);
                    pd.dismiss();
                    w = 0;
                    h = 13;
                }
            }
            super.handleMessage(message);
        }
    }

    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        setContentView(R.layout.main);
        this.pb = findViewById(R.id.progress);
        this.path = findViewById(R.id.path);
        this.find = findViewById(R.id.find);
        this.screen = findViewById(R.id.screen);
        this.setting = findViewById(R.id.setting);
        this.make = findViewById(R.id.make);
        OnClickListener anonymousClass1000000013 = new OnClickListener() {
            public void onClick(View view) {
                try {
                    switch (view.getId()) {
                        case R.id.find:
                            FileChooserDialog fileChooserDialog = new FileChooserDialog(MainActivity.this);
                            FileChooserDialog fileChooserDialog3 = fileChooserDialog;
                            fileChooserDialog3.setFilter(".*TTF|.*ttf");
                            fileChooserDialog = fileChooserDialog3;
                            fileChooserDialog.addListener(new OnFileSelectedListener() {
                                public void onFileSelected(Dialog dialog, File file, String str) {
                                }

                                public void onFileSelected(Dialog dialog, File file) {
                                    path.setText(file.getAbsolutePath());
                                    tf = Typeface.createFromFile(file);
                                    dialog.hide();
                                }
                            });
                            fileChooserDialog3.show();
                            return;
                        case R.id.setting:
                            showSettingDialog();
                            return;
                        case R.id.make:
                            pd = new ProgressDialog(MainActivity.this);
                            pd.setProgressStyle(1);
                            pd.setTitle("입력중입니다...");
                            pd.setMessage("Wait");
                            pd.setMax(11172);
                            pd.show();
                            makeFont(size);
                            setting.setClickable(false);
                            make.setClickable(false);
                            return;
                        default:
                    }
                } catch (Exception e) {
                    StringBuilder stringBuffer = new StringBuilder();
                    Log.i(MainActivity.TAG, stringBuffer.append(e).append("").toString());
                }
            }
        };
        this.make.setOnClickListener(anonymousClass1000000013);
        this.find.setOnClickListener(anonymousClass1000000013);
        this.setting.setOnClickListener(anonymousClass1000000013);
    }

    public void showSettingDialog() {
        this.prex = this.w;
        this.prey = this.h - 13;
        this.presize = this.size;
        View inflate = View.inflate(this, R.layout.setting, null);
        ImageView imageView = inflate.findViewById(R.id.screen2);
        this.sizes = inflate.findViewById(R.id.sizes);
        this.xs = inflate.findViewById(R.id.xs);
        this.ys = inflate.findViewById(R.id.ys);
        this.sizes.setProgress(this.size);
        this.xs.setProgress(this.w + 16);
        this.ys.setProgress(this.h + 3);
        imageView.setImageBitmap(getSample(this.w, this.h, this.size));
        AnonymousClass100000002 anonymousClass100000002 = new AnonymousClass100000002(this, imageView);
        this.sizes.setOnSeekBarChangeListener(anonymousClass100000002);
        this.xs.setOnSeekBarChangeListener(anonymousClass100000002);
        this.ys.setOnSeekBarChangeListener(anonymousClass100000002);
        Button button = inflate.findViewById(R.id.sizem);
        Button button2 = inflate.findViewById(R.id.sizep);
        Button button3 = inflate.findViewById(R.id.xm);
        Button button4 = inflate.findViewById(R.id.xp);
        Button button5 = inflate.findViewById(R.id.ym);
        Button button6 = inflate.findViewById(R.id.yp);
        AnonymousClass100000003 anonymousClass100000003 = new AnonymousClass100000003(this, imageView);
        button.setOnClickListener(anonymousClass100000003);
        button2.setOnClickListener(anonymousClass100000003);
        button3.setOnClickListener(anonymousClass100000003);
        button4.setOnClickListener(anonymousClass100000003);
        button5.setOnClickListener(anonymousClass100000003);
        button6.setOnClickListener(anonymousClass100000003);
        Builder builder = new Builder(this)
                .setView(inflate)
                .setTitle("스타일 설정하기")
                .setPositiveButton("확인", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialogInterface, int i) {
                        w = prex;
                        h = prey + 13;
                        size = presize;
                    }
                })
                .setNegativeButton("취소", null);
        builder.create().show();
    }

    public void makeFont(int i) {
        AnonymousClass100000005 anonymousClass100000005 = new AnonymousClass100000005(this, i);
        Thread thread = new Thread(anonymousClass100000005);
        thread.start();
    }

    public Bitmap getSample(int i, int i2, int i3) {
        Bitmap createBitmap = Bitmap.createBitmap(16, 16, Config.ARGB_8888);
        createBitmap.eraseColor(-1);
        Canvas canvas = new Canvas(createBitmap);
        Paint paint = new Paint();
        if (this.tf != null) {
            Typeface typeface = paint.setTypeface(tf);
        }
        paint.setColor(-16777216);
        paint.setTextSize((float) i3);
        paint.setAntiAlias(false);
        canvas.drawText("가", (float) i, (float) i2, paint);
        return Bitmap.createScaledBitmap(createBitmap, 256, 256, false);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        Menu menu2 = menu;
        MenuItem add = menu2.add(1, 123, 1, "정보");
        return super.onCreateOptionsMenu(menu2);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        Context context = this;
        if (menuItem.getItemId() == 123) {
            Intent intent;
            try {
                intent = new Intent(context, Class.forName("ifteam.affogatoman.fontgen.LicenseActivity"));
                startActivity(intent);
            } catch (Throwable e) {
                throw new NoClassDefFoundError(e.getMessage());
            }
        }
        return super.onOptionsItemSelected(menuItem);
    }

    public MainActivity() {
        this.canvas = new Canvas(this.base);
        this.p = new Paint();
        this.handler = new MyHandler();
        this.size = 16;
        this.w = 0;
        this.h = 13;
    }
}

File: app/src/main/java/ifteam/affogatoman/fontgen/MainActivity.kt
package ifteam.affogatoman.fontgen

import android.Manifest
import android.app.*
import android.content.*
import android.content.pm.PackageManager
import android.graphics.*
import android.os.*
import android.util.Log
import android.view.*
import android.widget.*
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.obsez.android.lib.filechooser.ChooserDialog
import kotlinx.android.synthetic.main.setting.*
import java.io.*
import kotlin.math.pow

class MainActivity : AppCompatActivity() {
    var baseBitmap: Bitmap = Bitmap.createBitmap(256, 256, Bitmap.Config.ARGB_8888)
    var canvas: Canvas
    var paint: Paint
    var typeFace: Typeface? = null

    var handler: MyHandler

    lateinit var findButton: Button
    lateinit var makeButton: Button
    lateinit var settingButton: Button
    lateinit var pathEditText: EditText
    lateinit var bitmapSizeSeekBar: SeekBar

    var screenImageView: ImageView? = null
    var characterProgressBar: ProgressBar? = null
    var characterProgressDialog: ProgressDialog? = null
    lateinit var sizeSeekBar: SeekBar
    lateinit var xSeekBar: SeekBar
    lateinit var ySeekBar: SeekBar
    lateinit var antiAliasSwitch: Switch
    lateinit var onlyKoreanSwitch: Switch

    var settingFontSize = 0
    var settingDrawX = 0
    var settingDrawY = 0

    var isAntiAlias: Boolean
    var isOnlyKorean: Boolean
    var fontSize: Int
    var drawPaddingX = 0
    var drawPaddingY = 0

    var bitmapSize: Int = 256

    internal inner class FontGenOnSeekBarChangeListener(private val mainActivity: MainActivity, private val settingImageView: ImageView?) : OnSeekBarChangeListener {
        override fun onStartTrackingTouch(seekBar: SeekBar) {}
        override fun onProgressChanged(seekBar: SeekBar, progress: Int, z: Boolean) {
            when (seekBar.id) {
                R.id.sizes ->  mainActivity.settingFontSize = progress
                R.id.xs -> mainActivity.settingDrawX = progress - xSeekBar.max/2
                R.id.ys -> mainActivity.settingDrawY = progress - ySeekBar.max/2
                R.id.ys -> mainActivity.settingDrawY = progress - ySeekBar.max/2
                R.id.size_seekbar -> {
                    val tempSize = bitmapSize
                    mainActivity.bitmapSize = 2.0.pow(progress + 8).toInt()
                    mainActivity.fontSize = mainActivity.bitmapSize/16
                    mainActivity.drawPaddingX = drawPaddingX * bitmapSize / tempSize
                    mainActivity.drawPaddingY = drawPaddingY * bitmapSize / tempSize
                    mainActivity.findViewById<TextView>(R.id.size_text).text = "$bitmapSize × $bitmapSize"
                }
                else -> { }
            }
            settingImageView?.setImageBitmap(mainActivity.getSample(mainActivity.settingDrawX, mainActivity.settingDrawY, mainActivity.settingFontSize))
        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {
            settingImageView?.setImageBitmap(mainActivity.getSample(mainActivity.settingDrawX, mainActivity.settingDrawY, mainActivity.settingFontSize))
        }

    }

    internal class FontGenOnClickListener(private val mainActivity: MainActivity, private val sampleImageView: ImageView) : View.OnClickListener {
        override fun onClick(view: View) {
            when (view.id) {
                R.id.sizem -> {
                    if (this.mainActivity.sizeSeekBar.progress != 0) {
                        this.mainActivity.settingFontSize--
                        this.mainActivity.sizeSeekBar.incrementProgressBy(-1)
                    }
                }
                R.id.sizep -> {
                    if (this.mainActivity.sizeSeekBar.progress != this.mainActivity.sizeSeekBar.max) {
                        this.mainActivity.settingFontSize++
                        this.mainActivity.sizeSeekBar.incrementProgressBy(1)
                    }
                }
                R.id.xm -> {
                    if (this.mainActivity.xSeekBar.progress != 0) {
                        this.mainActivity.settingDrawX--
                        this.mainActivity.xSeekBar.incrementProgressBy(-1)
                    }
                }
                R.id.xp -> {
                    if (this.mainActivity.xSeekBar.progress != this.mainActivity.xSeekBar.max) {
                        this.mainActivity.settingDrawX++
                        this.mainActivity.xSeekBar.incrementProgressBy(1)
                    }
                }
                R.id.ym -> {
                    if (this.mainActivity.ySeekBar.progress != 0) {
                        this.mainActivity.settingDrawY--
                        this.mainActivity.ySeekBar.incrementProgressBy(-1)
                    }
                }
                R.id.yp -> {
                    if (this.mainActivity.ySeekBar.progress != this.mainActivity.ySeekBar.max) {
                        this.mainActivity.settingDrawY++
                        this.mainActivity.ySeekBar.incrementProgressBy(1)
                    }
                }
                R.id.anti_alias -> this.mainActivity.isAntiAlias = (view as Switch).isChecked
                R.id.only_korean -> this.mainActivity.isOnlyKorean = (view as Switch).isChecked
                else -> { }
            }
            sampleImageView.setImageBitmap(mainActivity.getSample(mainActivity.settingDrawX, mainActivity.settingDrawY, mainActivity.settingFontSize))
        }

    }

    internal inner class FontGenRunnable(private val mainActivity: MainActivity, private val fontSize_: Int) : Runnable {
        var current = System.currentTimeMillis()
        var glyph = 0xAC
        var maxCount = '힣'.toInt()
        var startChar = '가'.toInt()
        var i = 44032

        override fun run() {
            try {
                var startX = drawPaddingX
                var startY = drawPaddingY

                if (!this.mainActivity.isOnlyKorean) {
                    glyph = 0x00
                    maxCount = 256*256-1
                    i = 0
                    startChar = 0
                }
                this.mainActivity.baseBitmap = Bitmap.createBitmap(bitmapSize, bitmapSize, Bitmap.Config.ARGB_8888)
                this.mainActivity.canvas = Canvas(this.mainActivity.baseBitmap)
                if (this.mainActivity.typeFace != null)
                    this.mainActivity.paint.typeface = this.mainActivity.typeFace
                this.mainActivity.paint.color = Color.WHITE
                this.mainActivity.paint.textSize = fontSize_.toFloat()
                this.mainActivity.paint.isAntiAlias = this.mainActivity.isAntiAlias

                val piece: Bitmap = Bitmap.createBitmap(this.mainActivity.bitmapSize/16, this.mainActivity.bitmapSize, Bitmap.Config.ARGB_8888)
                var pieceCanvas = Canvas(piece)
                val rect = Rect()

                while (i <= maxCount) {
                    var file: File
                    //Thread.sleep(1)
                    this.mainActivity.paint.getTextBounds(i.toChar().toString(), 0, 1, rect)
                    Log.i(TAG, rect.left.toString())
                    pieceCanvas.drawText(i.toChar().toString(), -rect.left.toFloat(), -rect.top.toFloat(), this.mainActivity.paint)
                    this.mainActivity.canvas.drawBitmap(piece, startX.toFloat(), startY.toFloat(), null)
                    piece.eraseColor(Color.TRANSPARENT)
                    startX += bitmapSize/16

                    val myHandler = this.mainActivity.handler
                    myHandler.sendMessage(myHandler.obtainMessage(100, arrayOf(i, maxCount, startChar)))

                    if (i == startChar+1) {
                        file = File("${Environment.getExternalStorageDirectory().absolutePath}/FontGen/$current/.nomedia")
                        file.parentFile.mkdirs()
                        file.createNewFile()
                    }
                    if ((i - startChar) % 16 == 15 && i != startChar) {
                        startX = drawPaddingX
                        startY += bitmapSize/16
                    }
                    if (startY - drawPaddingY >= bitmapSize || i == maxCount) {
                        file = File("${Environment.getExternalStorageDirectory().absolutePath}/FontGen/$current/glyph_"+"%02X".format(glyph)+".png")
                        file.parentFile.mkdirs()
                        val fileOutputStream: OutputStream = FileOutputStream(file)
                        val bufferedOutputStream: OutputStream = BufferedOutputStream(fileOutputStream)
                        this.mainActivity.baseBitmap.compress(Bitmap.CompressFormat.PNG, 100, bufferedOutputStream)
                        bufferedOutputStream.close()
                        fileOutputStream.close()
                        if (i != maxCount) {
                            this.mainActivity.baseBitmap.eraseColor(Color.TRANSPARENT)
                        }
                        startY = drawPaddingY
                        glyph++
                    }
                    i++
                }
            } catch (e: IOException) {
                Log.i(TAG, e.toString())
            }
        }

    }

    inner class MyHandler : Handler() {
        override fun handleMessage(message: Message) {
            if (message.what == 100) {
                makeButton.text = (message.obj as Array<Int>)[0].toChar().toString()
                if(((message.obj as Array<Int>)[0]-(message.obj as Array<Int>)[2]) % 64 == 63 || (message.obj as Array<Int>)[0] == (message.obj as Array<Int>)[1])
                    screenImageView!!.setImageBitmap(baseBitmap)
                characterProgressDialog!!.setMessage(getString(R.string.current_char)+(message.obj as Array<Int>)[0].toChar().toString())
                characterProgressBar!!.max = (message.obj as Array<Int>)[1]-(message.obj as Array<Int>)[2]
                characterProgressBar!!.progress = (message.obj as Array<Int>)[0]-(message.obj as Array<Int>)[2]
                characterProgressDialog!!.max = (message.obj as Array<Int>)[1]-(message.obj as Array<Int>)[2]
                characterProgressDialog!!.progress = (message.obj as Array<Int>)[0]-(message.obj as Array<Int>)[2]
                if ((message.obj as Array<Int>)[0] == (message.obj as Array<Int>)[1]) {
                    makeButton.text = getString(R.string.make)
                    makeButton.isEnabled = true
                    settingButton.isEnabled = true
                    characterProgressBar!!.progress = 0
                    characterProgressDialog!!.dismiss()
                }
            }
            super.handleMessage(message)
        }
    }

    public override fun onCreate(bundle: Bundle?) {
        super.onCreate(bundle)

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED)
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 0)

        setContentView(R.layout.main)
        bitmapSizeSeekBar = findViewById(R.id.size_seekbar)
        characterProgressBar = findViewById(R.id.progress)
        pathEditText = findViewById(R.id.path)
        findButton = findViewById(R.id.find)
        screenImageView = findViewById(R.id.screen)
        settingButton = findViewById(R.id.setting)
        makeButton = findViewById(R.id.make)
        val viewOnClickListener = View.OnClickListener { view ->
            try {
                when (view.id) {
                    R.id.find -> {
                        val chooserDialog = ChooserDialog(this@MainActivity)
                        chooserDialog.withOnBackPressedListener { _ -> chooserDialog.goBack() }
                                .withOnLastBackPressedListener { dialog -> dialog.cancel() }
                                .withFilter(false, true, "ttf", "otf")
                                .withResources(R.string.choose_ttf_file, R.string.choose_ttf_file, R.string.cancel)
                                .withChosenListener { _, file ->
                                    pathEditText.setText(file.absolutePath)
                                    typeFace = Typeface.createFromFile(file)
                                }
                                .build()
                                .show()
                    }
                    R.id.setting -> showSettingDialog()
                    R.id.make -> {
                        characterProgressDialog = ProgressDialog(this@MainActivity)
                        characterProgressDialog!!.setProgressStyle(1)
                        characterProgressDialog!!.setTitle(R.string.writing)
                        characterProgressDialog!!.setMessage("Wait")
                        characterProgressDialog!!.max = 11172
                        characterProgressDialog!!.show()
                        makeFont(fontSize)
                        settingButton.isEnabled = false
                        makeButton.isEnabled = false
                    }
                    else -> return@OnClickListener
                }
            } catch (e: Exception) {
                val stringBuffer = StringBuilder()
                Log.i(TAG, stringBuffer.append(e).append("").toString())
            }
        }
        makeButton.setOnClickListener(viewOnClickListener)
        findButton.setOnClickListener(viewOnClickListener)
        settingButton.setOnClickListener(viewOnClickListener)
        bitmapSizeSeekBar.setOnSeekBarChangeListener(FontGenOnSeekBarChangeListener(this, null))
    }

    private fun showSettingDialog() {
        val tempIsAntiAlias: Boolean = isAntiAlias
        val tempIsOnlyKorean: Boolean = isOnlyKorean
        settingDrawX = drawPaddingX
        settingDrawY = drawPaddingY
        settingFontSize = fontSize
        val inflate = View.inflate(this, R.layout.setting, null)
        val imageView = inflate.findViewById<ImageView>(R.id.screen2)
        sizeSeekBar = inflate.findViewById(R.id.sizes)
        xSeekBar = inflate.findViewById(R.id.xs)
        ySeekBar = inflate.findViewById(R.id.ys)

        sizeSeekBar.max = bitmapSize/16
        xSeekBar.max = bitmapSize/8
        ySeekBar.max = bitmapSize/8

        antiAliasSwitch = inflate.findViewById(R.id.anti_alias)
        antiAliasSwitch.isChecked = isAntiAlias
        onlyKoreanSwitch = inflate.findViewById(R.id.only_korean)
        onlyKoreanSwitch.isChecked = isOnlyKorean

        sizeSeekBar.progress = fontSize
        xSeekBar.progress = drawPaddingX + (16*bitmapSize/256)
        ySeekBar.progress = drawPaddingY + (16*bitmapSize/256)

        imageView.setImageBitmap(getSample(drawPaddingX, drawPaddingY, fontSize))
        val fontGenOnSeekBarChangeListener = FontGenOnSeekBarChangeListener(this, imageView)
        sizeSeekBar.setOnSeekBarChangeListener(fontGenOnSeekBarChangeListener)
        xSeekBar.setOnSeekBarChangeListener(fontGenOnSeekBarChangeListener)
        ySeekBar.setOnSeekBarChangeListener(fontGenOnSeekBarChangeListener)
        val fontGenOnClickListener = FontGenOnClickListener(this, imageView)
        inflate.findViewById<Button>(R.id.sizem).setOnClickListener(fontGenOnClickListener)
        inflate.findViewById<Button>(R.id.sizep).setOnClickListener(fontGenOnClickListener)
        inflate.findViewById<Button>(R.id.xm).setOnClickListener(fontGenOnClickListener)
        inflate.findViewById<Button>(R.id.xp).setOnClickListener(fontGenOnClickListener)
        inflate.findViewById<Button>(R.id.ym).setOnClickListener(fontGenOnClickListener)
        inflate.findViewById<Button>(R.id.yp).setOnClickListener(fontGenOnClickListener)
        antiAliasSwitch.setOnClickListener(fontGenOnClickListener)
        onlyKoreanSwitch.setOnClickListener(fontGenOnClickListener)
        val builder = AlertDialog.Builder(this)
                .setView(inflate)
                .setTitle(R.string.setting_title)
                .setPositiveButton(R.string.confirm) { _, _ ->
                    drawPaddingX = settingDrawX
                    drawPaddingY = settingDrawY
                    fontSize = settingFontSize
                }
                .setNegativeButton(R.string.cancel) { _, _ ->
                    isAntiAlias = tempIsAntiAlias
                    isOnlyKorean = tempIsOnlyKorean
                }
        builder.create().show()
    }

    private fun makeFont(_fontSize: Int) {
        val fontGenRunnable = FontGenRunnable(this, _fontSize)
        val thread = Thread(fontGenRunnable)
        thread.start()
    }

    fun getSample(paddingX: Int, paddingY: Int, fontSize: Int): Bitmap {
        val createBitmap = Bitmap.createBitmap(bitmapSize/16, bitmapSize/16, Bitmap.Config.ARGB_8888)
        createBitmap.eraseColor(Color.WHITE)
        val canvas = Canvas(createBitmap)
        val paint_ = Paint()
        if (typeFace != null)
            paint_.typeface = typeFace
        paint_.color = Color.BLACK
        paint_.textSize = fontSize.toFloat()
        paint_.isAntiAlias = isAntiAlias
        val rect = Rect()
        paint_.getTextBounds("가", 0, 1, rect)
        canvas.drawText("가", -rect.left.toFloat()+paddingX, -rect.top.toFloat()+paddingY, paint_)
        return Bitmap.createScaledBitmap(createBitmap, 256, 256, false)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val add = menu.add(1, 123, 1, R.string.info)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        if (menuItem.itemId == 123) {
            try {
                val intent = Intent(this, Class.forName("ifteam.affogatoman.fontgen.LicenseActivity"))
                startActivity(intent)
            } catch (e: Throwable) {
                throw NoClassDefFoundError(e.message)
            }
        }
        return super.onOptionsItemSelected(menuItem)
    }

    companion object {
        const val TAG = "AFFO"
    }

    init {
        canvas = Canvas(baseBitmap)
        paint = Paint()
        handler = MyHandler()
        fontSize = bitmapSize/16
        isAntiAlias = true
        isOnlyKorean = true
    }
}